momento-provider/src/main/java/com/google/code/ssm/providers/momento/MomentoClientWrapper.java

```diff
@@ -217,11 +196,67 @@ public Object getNativeClient() {
-                | (data[i + 2] & 0xff) << 8
-                | (data[i + 3] & 0xff);
+    // Helper function that concats the CachedObject to raw bytes before calling `set`.
+    // This helps maintain our serialization flags so that we can deserialize the object upon a `get`
+    private boolean writeOutToMomento(final String key, final int exp, final int flagsUsed, final byte[] data) {
+        final byte[] flags = intToBytes(flagsUsed);
+        ByteBuffer buffer = ByteBuffer.allocate(flags.length + data.length);
+        buffer.put(flags);
+        buffer.put(data);
+        buffer.rewind();
+        CacheSetResponse response = momentoClient.set(defaultCacheName, key, buffer.asReadOnlyBuffer(), exp);
+        return response != null;
+    }
+
+    // Helper function that reads the raw bytes from Momento and splits up the bytes
+    // to extract the serialization flags we used as well as the raw bytes of data that contain our
+    // actual object
+    private <T> T readFromMomento(final String key, final CacheTranscoder transcoder) {
+        Transcoder<T> cacheTranscoder = getTranscoder(transcoder);
+        CacheGetResponse response = momentoClient.get(defaultCacheName, key);
+        Optional<byte[]> cacheGetResponse = response.byteArray();
+        if (cacheGetResponse.isPresent()) {
+            byte[] returnedBytes = cacheGetResponse.get();
+            int flags = getFlags(returnedBytes);
+            byte[] originalData = getData(returnedBytes);
+            return cacheTranscoder.decode(new CachedData(flags, originalData, originalData.length));
+        }
+        return null;
+    }
+
+    // Helper function that reads the raw bytes from Momento and splits up the bytes
+    // to extract the serialization flags we used as well as the raw bytes of data that contain our
+    // actual object, uses our default Transcoder
+    private Object readFromMomento(final String key) {
+        CacheTranscoder cacheTranscoder = getTranscoder();
+        CacheGetResponse response = momentoClient.get(defaultCacheName, key);
+        Optional<byte[]> cacheGetResponse = response.byteArray();
+        if (cacheGetResponse.isPresent()) {
+            byte[] returnedBytes = cacheGetResponse.get();
+            int flags = getFlags(returnedBytes);
+            byte[] originalData = getData(returnedBytes);
+            return cacheTranscoder.decode(new CachedObjectWrapper(new CachedData(flags, originalData, originalData.length)));
+        }
+        return null;
+    }
+
+    private static byte[] intToBytes(final int i) {
+        ByteBuffer byteBuffer = ByteBuffer.allocate(EXTRA_HDR_LEN);
+        byteBuffer.putInt(i);
+        return byteBuffer.array();
+    }
+
+    private static int getFlags(byte[] data) {
+        byte[] flags = Arrays.copyOfRange(data, 0, EXTRA_HDR_LEN);
+        int flag = 0;
+        for (byte b: flags) {
+            flag = (flag << 8) + (b & 0xFF);
+        }
+        return flag;
+    }
```

Comment:
So, compared to the previous `getFlag()` this has:

* a built-in dependency on en external `EXTRA_HDR_LEN` which needs to definitely be 4.
* a built-in dependency on ByteBuffer's `putInt()` implementation assuming that this is exactly the reflexive operation.
* a byte array heap allocation to read a 4 bytes integer.
* a loop construct hiding "4" from $future_dev, who will know that `int flag` can only hold 4 and may be misled that EXTRA_HDR_LEN could possibly vary so they'd need to check (I did)

```
➜  simple-spring-memcached git:(momento) grep -r EXTRA_HDR_LEN .
./momento-provider/src/main/java/com/google/code/ssm/providers/momento/MomentoClientWrapper.java:    private static final int EXTRA_HDR_LEN = 4;
```

I think based on your implementation of intToBytes what you want here is
```
private static int getFlags(byte[] data) {
    return ByteBuffer.wrap(data).getInt();
}
```

edit: and based on the rest of the comments, I think you want to just remove this method and replace it with `const int flags = responseBuffer.getInt();`

Reply:
> a built-in dependency on en external EXTRA_HDR_LEN which needs to definitely be 4.

This isn't external, this is defined in the same file and is most definitely 4, I just kept the name the same. Happy to remove it for clarity!

But yes, we should be able to just replace this with `repsonseBuffer.getInt()`